File: cpp/topological_sort.cpp
vector<int> topological_sort(vector<vector<int>>& to){
    int n = to.size();
    vector<int> count(n);

    for(int i = 0;i < n; ++i){
        for(int v: to[i]){
            count[v]++;
        }
    }

    queue<int> zero;
    for(int i = 0;i < n; ++i){
        if(count[i] == 0)zero.push(i);
    }

    vector<int> sorted;
    while(!zero.empty()){
        int v = zero.front();
        zero.pop();
        for(int u : to[v]){
            count[u]--;
            if(count[u] == 0)zero.push(u);
        }
        sorted.push_back(v);
    }

    return sorted;
}

File: rust/topological_sort.rs
fn topological_sort(to: &Vec<Vec<usize>>) -> Vec<usize> {
    let n = to.len();
    let mut cnt = vec![0; n];
    for i in 0..n {
        for &v in &to[i] {
            cnt[v] += 1;
        }
    }

    let mut zero = VecDeque::new();
    for i in 0..n {
        if cnt[i] == 0 {
            zero.push_back(i);
        }
    }

    let mut sorted = vec![];
    while let Some(v) = zero.pop_front() {
        sorted.push(v);
        for &u in &to[v] {
            cnt[u] -= 1;
            if cnt[u] == 0 {
                zero.push_back(u);
            }
        }
    }

    sorted
}
